### PR TITLE
Add doc for using GPU in Knative

### PR DESCRIPTION
**Pull request by SiyangShao**
## Summary

Add doc for using GPU in Knative

## Implementation Notes :hammer_and_pick:

* adding scripts and document, in order to enable GPU in stock Knative which is installed using current scripts.

## External Dependencies :four_leaf_clover:

* 

## Breaking API Changes :warning:

* 

*Simply specify none (N/A) if not applicable.*

### CODE REVIEW

File: docs/developers_guide.md [hunk 1]
```diff
@@ -219,3 +219,109 @@ We plan to keep our code loosely up to date with the upstream Firecracker reposi
...
+./scripts/gpu/setup_nvidiadevice.sh
+```
+At this point, all pods should be successfully deployed.
+``` bash
+$ kubectl get pods -A
+NAMESPACE              NAME                                       READY   STATUS      RESTARTS   AGE
+nvidia-device-plugin   nvidia-device-plugin-1684892866-rtgk9      1/1     Running     0          80s
+
+```
+And you can use the gpu-pod.yaml to test whether the Nvidia Device Plugin works or not
```
Comment (ustiugov) [inline docs/developers_guide.md:273]:
```suggestion
And you can use the gpu-pod.yaml to test whether the Nvidia Device Plugin works
```

File: docs/developers_guide.md [hunk 2]
```diff
@@ -219,3 +219,109 @@ We plan to keep our code loosely up to date with the upstream Firecracker reposi
...
+
+``` bash
+kubectl apply -f ./scripts/gpu/gpu-pod.yaml
+```
+After it completed, check the log and it should be
```
Comment (ustiugov) [inline docs/developers_guide.md:278]:
```suggestion
After that, check the log to contain:
```

File: docs/developers_guide.md [hunk 3]
```diff
@@ -219,3 +219,109 @@ We plan to keep our code loosely up to date with the upstream Firecracker reposi
...
+``` bash
+[Vector addition of 50000 elements]
+Copy input data from the host memory to the CUDA device
+CUDA kernel launch with 196 blocks of 256 threads
+Copy output data from the CUDA device to the host memory
+Test PASSED
+Done
+```
+### Test Knative with GPU
+Using GPU in Knative is simple and same to the method in Kubernetes. The only change is to add the limits of GPU under the containers part.
```
Comment (ustiugov) [inline docs/developers_guide.md:291]:
```suggestion
Using GPU in Knative is simple and similar as for a regular Kubernetes service. The only change is to add the limits of GPU to the YAML file.
```

File: docs/developers_guide.md [hunk 4]
```diff
@@ -219,3 +219,109 @@ We plan to keep our code loosely up to date with the upstream Firecracker reposi
...
+``` yaml
+resources:
+  limits:
+    nvidia.com/gpu: 1
+```
+You can also try deploying our container, which is a simple Go service that calls and returns the output of nvidia-smi.
```
Comment (ustiugov) [inline docs/developers_guide.md:297]:
```suggestion
You can also deploy our example container, which is a Golang function that calls and returns the output of nvidia-smi.
```

File: docs/developers_guide.md [hunk 5]
```diff
@@ -219,3 +219,109 @@ We plan to keep our code loosely up to date with the upstream Firecracker reposi
...
+``` bash
+kn service apply -f ./scripts/gpu/gpu-service.yaml
```
Comment (ustiugov) [inline docs/developers_guide.md:299]: is it a service or a Knative function? we need only the function example

File: docs/developers_guide.md [hunk 6]
```diff
@@ -219,3 +219,109 @@ We plan to keep our code loosely up to date with the upstream Firecracker reposi
...
+| N/A   35C    P0               27W / 250W|      4MiB / 16384MiB |      0%      Default |
+|                                         |                      |                  N/A |
++-----------------------------------------+----------------------+----------------------+
+                                                                                         
++---------------------------------------------------------------------------------------+
+| Processes:                                                                            |
+|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
+|        ID   ID                                                             Usage      |
+|=======================================================================================|
++---------------------------------------------------------------------------------------+
```
Comment (ustiugov) [inline docs/developers_guide.md:325]: enough to show the header of the output, no need to show a full output.

File: scripts/gpu/setup_nvidiacontainerd.sh [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+distribution=$(. /etc/os-release;echo $ID$VERSION_ID) \
+   && curl -s -L https://nvidia.github.io/nvidia-docker/gpgkey | sudo apt-key add - \
+   && curl -s -L https://nvidia.github.io/nvidia-docker/$distribution/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list
+
+sudo apt-get update \
+   && sudo apt-get install -y nvidia-container-runtime
+
+# Create default containerd and manually modify it, based on the nvidia document
+# sudo mkdir -p /etc/containerd \
+#    && sudo containerd config default | sudo tee /etc/containerd/config.toml
```
Comment (ustiugov) [inline scripts/gpu/setup_nvidiacontainerd.sh:16]: clean up

File: scripts/gpu/setup_nvidiadevice.sh [hunk 8]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+# Follow the Nvidia guide, use helm to install it
+curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3 \
+   && chmod 700 get_helm.sh \
+   && ./get_helm.sh
+
+helm repo add nvdp https://nvidia.github.io/k8s-device-plugin \
+   && helm repo update
```
Comment (ustiugov) [inline scripts/gpu/setup_nvidiadevice.sh:9]: please move this code to setup_system.sh, making `helm` one of the vHive dependencies.

File: scripts/gpu/setup_cudadriver.sh [hunk 9]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+sudo apt-get install linux-headers-$(uname -r)
+distribution=$(. /etc/os-release;echo $ID$VERSION_ID | sed -e 's/\.//g') \
+   && wget https://developer.download.nvidia.com/compute/cuda/repos/$distribution/x86_64/cuda-$distribution.pin \
+   && sudo mv cuda-$distribution.pin /etc/apt/preferences.d/cuda-repository-pin-600
```
Comment (ustiugov) [inline scripts/gpu/setup_cudadriver.sh:6]: where is $distribution defined?

**Review by ustiugov — CHANGES_REQUESTED** [review]
thank you! 

Todo:

- [ ] move yaml files to the `configs/gpu`
- [ ] instead of having containerd_config.toml in our repo, use `wget` to fetch it from its origin. 
- [ ] add vHive copyrights similarly to other files in vHive
- [ ] merge setup_[driver|containerd] into a single script to reduce the number of files
- [ ] apply other comments which i added to the code
- [ ] ask Ruiqi to try out the manual
- [ ] add a note to the CHANGELOG

File: docs/developers_guide.md (hunk 5 above)
Comment (SiyangShao) [reply docs/developers_guide.md:299]: It's a service. For Knative function version, it needs extra dependency for kn func. I'll try to change it to a function example

File: scripts/gpu/setup_nvidiadevice.sh (hunk 8 above)
Comment (SiyangShao) [reply scripts/gpu/setup_nvidiadevice.sh:9]: Do we need specify whether NVIDIA is needed for that? As helm is only used for NVIDIA

File: scripts/gpu/setup_cudadriver.sh (hunk 9 above)
Comment (SiyangShao) [reply scripts/gpu/setup_cudadriver.sh:6]: It's definition is just
`distribution=$(. /etc/os-release;echo $ID$VERSION_ID | sed -e 's/\.//g')`
The code here is just to check system version, i.e., ubuntu2004 or ubuntu2204.

**Review by ustiugov — CHANGES_REQUESTED** [review]
more comments

**Review by ustiugov — CHANGES_REQUESTED** [review]
ok but need to replace the service example with a function.

**Comment by ustiugov** [comment]
also, please fix the linter errors.

File: configs/gpu/gpu-function.yaml [hunk 10]
```diff
@@ -0,0 +1,17 @@
+apiVersion: serving.knative.dev/v1
+kind: Service
+metadata:
+  name: hello-gpu
+spec:
+  template:
+    spec:
+      containers:
+        - image: ghcr.io/vhive-serverless/hello-gpu:latest
```
Comment (ustiugov) [inline configs/gpu/gpu-function.yaml:9]: can you please add the source code and the dockerfile to `function-images/`? or leave the link to the source code in a comment

File: docs/developers_guide.md [hunk 11]
```diff
@@ -219,3 +219,88 @@ We plan to keep our code loosely up to date with the upstream Firecracker reposi
...
+### Install Stock Kubernetes and Knative
+
+Follow the guide to [setup stock knative](#testing-stock-knative-setup-or-images). 
+
+``` bash
+./scripts/cloudlab/setup_node.sh stock-only
+```
+
+### Install NVIDIA Driver and NVIDIA Container Toolkit 
+The script will install NVIDIA CUDA Driver and assume there’s no NVIDIA driver currently running. You can use the script provided if the install of containerd is using our script or manually edit the containerd settings following [NVIDIA's official document](https://docs.nvidia.com/datacenter/cloud-native/kubernetes/install-k8s.html#install-nvidia-container-toolkit-nvidia-docker2).
```
Comment (ustiugov) [inline docs/developers_guide.md:237]: please specify the specs of the platform, i.e., CUDA and GPU version, on which you tested the following commands. Give recommendations on which platforms/GPUs the commands should work. 

split the long line into several lines for finer-grain version control

File: docs/developers_guide.md [hunk 12]
```diff
@@ -219,3 +219,88 @@ We plan to keep our code loosely up to date with the upstream Firecracker reposi
...
+    nvidia.com/gpu: 1
+```
+You can also deploy our example container, which is a Golang function that calls and returns the output of `nvidia-smi`.
+
+``` bash
+kn service apply -f ./configs/gpu/gpu-function.yaml
+```
+Once the service has been deployed successfully, you can call it and check it's response.
+``` bash
+curl "$(kn service describe hello-gpu -o url)"
```
Comment (ustiugov) [inline docs/developers_guide.md:297]:
```suggestion
curl "$(kn service describe hello-gpu -o <URL>)"
```

File: scripts/gpu/setup_nvidia.sh [hunk 13]
```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
```
Comment (ustiugov) [inline scripts/gpu/setup_nvidia.sh:1]: rename to `setup_nvidia_gpu.sh`

**Review by ustiugov — CHANGES_REQUESTED** [review]
few minor changes

**Review by ustiugov — CHANGES_REQUESTED** [review]
add our copyrights to the source files of the GPU container (if it's your code) or the original copyrights (if it's not)

**Review by ustiugov — APPROVED** [review]

**Review by ustiugov — APPROVED** [review]

**Review by ustiugov — APPROVED** [review]

**Pull request merged**